mono/metadata/ee.c

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed to the .NET Foundation under one or more agreements.
+ * The .NET Foundation licenses this file to you under the MIT license.
+ * See the LICENSE file in the project root for more information.
+ */
+
+#include <config.h>
+#include <mono/metadata/ee.h>
+#include <mono/utils/mono-dl.h>
+#include <mono/eglib/glib.h>
+
+#define LIB_NAME "mini-ee-%s"
+#define INITIALIZER_NAME "mono_ee_%s_init"
+#define VERSION_NAME "mono_ee_%s_version"
+
+typedef int (*MonoEEVersion) (void);
+typedef void (*MonoEEInit) (const char*);
+
+static gboolean
+load_ee (MonoDl *module, const char *name, const char *opts)
+{
+	if (!module)
+		return FALSE;
+
+	char *err;
+	char *version_name = g_strdup_printf (VERSION_NAME, name);
+	MonoEEVersion ee_version;
+
+	if ((err = mono_dl_symbol (module, version_name, (gpointer *) &ee_version))) {
+		g_printerr ("Could not load symbol: %s\n", err);
+		g_free (err);
+		return FALSE;
+	}
+
+	int ee_module_version = ee_version ();
+	if (ee_module_version != MONO_EE_API_VERSION) {
+		g_printerr ("version mismatch for execution engine module %s: It's %d, but %d is required\n", name, ee_module_version, MONO_EE_API_VERSION);
+		return FALSE;
+	}
+
+	char *init_name = g_strdup_printf (INITIALIZER_NAME, name);
+	MonoEEInit func;
+
+	if ((err = mono_dl_symbol (module, init_name, (gpointer *) &func))) {
+		g_printerr ("Could not load symbol: %s\n", err);
+		g_free (err);
+		return FALSE;
+	}
+
+	func (opts);
+
+	return TRUE;
+}
+
+static gboolean
+load_ee_from_executable (const char *name, const char *opts)
+{
+	char *err;
+	MonoDl *module = mono_dl_open (NULL, MONO_DL_EAGER, &err);
```

Comment:
@spouliot @rolfbjarne just to make sure: is this going to work on iOS?

with the profiler we do have a similar problem, but it's solved by generating code in mtouch to initialize a profiler module. if `dlopen` etc. is discouraged, we can do something similar with the interpreter.

Reply:
I don't think we need to be able to load the interpreter dynamically only statically, i.e. if the interpreter is disabled, then the embedding app should be able to call a function in libmono-interpreter.a like mono_interp_register () or something, which would register the interpreter with the runtime.

Reply:
How about if  interpreter provides a function that instals itself against the runtime it was linked against?

Reply:
@kumpera that's already the case, the question is
1. what's the visibility of this function
2. who should call it

@vargaz it's not really dynamically (yet). right now it just looks up the symbol in it's own process, so it must be linked in. In the future I would like to move the JIT to the same kind of interface, along with other execution engines such as `il2cpp`, and then loading it dynamically is getting more interesting.